File: E-Commerce App Project (Tabbed)/Controllers/SaleViewController.m
//
//  SaleViewController.m
//  E-Commerce App Project (Tabbed)
//
//  Created by Rony Banik on 7/11/18.
//  Copyright © 2018 Rony Banik. All rights reserved.
//

#import "SaleViewController.h"
#import "Item.h"
#import "Items.h"
#import "SingleItemViewController.h"
#import "ShoppingCart.h"

@interface SaleViewController ()

@end

@implementation SaleViewController

@synthesize saleCollectionView;
@synthesize saleItemsList;

- (id)init
{
    if((self = [super init]))
    {
        NSArray *allItemsList = [Items sharedInstance].allItems;
        saleItemsList = [self returnSaleArray:allItemsList];
    }
    return self;
}
- (void)viewDidLoad {
    [super viewDidLoad];
    
    saleCollectionView.dataSource=self;
    saleCollectionView.delegate=self;
    
}

/*
#pragma mark - Navigation
// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/

-(NSArray *)returnSaleArray:(NSArray *)allItems{
    
    NSMutableArray *saleItems;
    saleItems = [allItems mutableCopy];
    
    Item* runningItem;
    
    for (int i=0; i<allItems.count; i++) {
         if ([[allItems[i] sale] isEqualToNumber:@0]) {
            runningItem=allItems[i];
            [saleItems removeObject:allItems[i]];
        }
    }
    return saleItems;
}
- (nonnull __kindof UICollectionViewCell *)collectionView:(nonnull UICollectionView *)collectionView cellForItemAtIndexPath:(nonnull NSIndexPath *)indexPath {
    
    NSArray *allItemsList = [Items sharedInstance].allItems;
    saleItemsList = [self returnSaleArray:allItemsList];
    NSObject *individualData=saleItemsList[indexPath.row];
    UICollectionViewCell *collectioViewCell = [collectionView dequeueReusableCellWithReuseIdentifier:@"sellCellIdentifier" forIndexPath:indexPath];

    UIImageView *itemImage = (UIImageView *)[collectioViewCell viewWithTag:601];
    UILabel *itemName = (UILabel *)[collectioViewCell viewWithTag:602];
    UILabel *itemPrice = (UILabel *)[collectioViewCell viewWithTag:603];

    NSURL *url = [NSURL URLWithString: [individualData valueForKey:@"photoURL"]];
    NSData *data = [NSData dataWithContentsOfURL:url];
    itemImage.image=[UIImage imageWithData:data];
    itemName.text=[individualData valueForKey:@"itemName"];
    itemPrice.text= [self showPrice:[[individualData valueForKey:@"price"]doubleValue]];

    return collectioViewCell;
}


- (NSInteger)collectionView:(nonnull UICollectionView *)collectionView numberOfItemsInSection:(NSInteger)section {
    NSArray *allItemsList = [Items sharedInstance].allItems;
    saleItemsList = [self returnSaleArray:allItemsList];
    return saleItemsList.count;
}

- (void)collectionView:(UICollectionView *)collectionView didSelectItemAtIndexPath:(NSIndexPath *)indexPath{

    SingleItemViewController *singleItemView =[self.storyboard instantiateViewControllerWithIdentifier:@"singleItemViewStoryBoardIdentifier"];

        NSObject *individualData= [saleItemsList objectAtIndex:indexPath.row];

        singleItemView.itemObjectReceived = [saleItemsList objectAtIndex:indexPath.row];

        NSURL *url = [NSURL URLWithString: [individualData valueForKey:@"photoURL"]];
        NSData *data = [NSData dataWithContentsOfURL:url];
        singleItemView.setItemImage = data;
        singleItemView.setItemName = [individualData valueForKey:@"itemName"];
        singleItemView.setItemCategory = [individualData valueForKey:@"itemCategory"];
        singleItemView.setItemID = [individualData valueForKey:@"ID"];
        singleItemView.setItemPrice = [individualData valueForKey:@"price"];
        singleItemView.setItemBrand = [individualData valueForKey:@"brand"];
        singleItemView.setItemQuality = [individualData valueForKey:@"quality"];



    [self.navigationController pushViewController:singleItemView animated:YES];
}
-(NSString *)showPrice:(double)price{
    NSString *tempPriceString = [[NSNumber numberWithDouble:price] stringValue];
    NSString *priceString = [NSString stringWithFormat:@"$%@",tempPriceString];
    return priceString;
}
@end

File: E-Commerce App Project (Tabbed)/Controllers/SaleViewController.swift
//
//  SaleViewController.swift
//  E-Commerce App Project (Tabbed)
//
//  Converted to Swift
//

import UIKit

class SaleViewController: UIViewController, UICollectionViewDataSource, UICollectionViewDelegate {
    
    @IBOutlet var saleCollectionView: UICollectionView!
    var saleItemsList: [Item] = []
    
    override init(nibName nibNameOrNil: String?, bundle nibBundleOrNil: Bundle?) {
        super.init(nibName: nibNameOrNil, bundle: nibBundleOrNil)
        let allItemsList = Items.sharedInstance.allItems
        saleItemsList = returnSaleArray(allItemsList)
    }
    
    required init?(coder: NSCoder) {
        super.init(coder: coder)
        let allItemsList = Items.sharedInstance.allItems
        saleItemsList = returnSaleArray(allItemsList)
    }
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        saleCollectionView.dataSource = self
        saleCollectionView.delegate = self
    }
    
    func returnSaleArray(_ allItems: [Item]) -> [Item] {
        return allItems.filter { item in
            if let sale = item.sale {
                return sale.intValue != 0
            }
            return false
        }
    }
    
    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
        let allItemsList = Items.sharedInstance.allItems
        saleItemsList = returnSaleArray(allItemsList)
        let individualData = saleItemsList[indexPath.row]
        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: "sellCellIdentifier", for: indexPath)
        
        if let itemImage = cell.viewWithTag(601) as? UIImageView,
           let itemName = cell.viewWithTag(602) as? UILabel,
           let itemPrice = cell.viewWithTag(603) as? UILabel {
            if let url = URL(string: individualData.photoURL),
               let data = try? Data(contentsOf: url) {
                itemImage.image = UIImage(data: data)
            }
            itemName.text = individualData.itemName
            itemPrice.text = showPrice(individualData.price)
        }
        
        return cell
    }
    
    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
        let allItemsList = Items.sharedInstance.allItems
        saleItemsList = returnSaleArray(allItemsList)
        return saleItemsList.count
    }
    
    func collectionView(_ collectionView: UICollectionView, didSelectItemAt indexPath: IndexPath) {
        guard let singleItemView = storyboard?.instantiateViewController(withIdentifier: "singleItemViewStoryBoardIdentifier") as? SingleItemViewController else { return }
        
        let individualData = saleItemsList[indexPath.row]
        
        singleItemView.itemObjectReceived = individualData
        
        if let url = URL(string: individualData.photoURL),
           let data = try? Data(contentsOf: url) {
            singleItemView.setItemImage = data
        }
        singleItemView.setItemName = individualData.itemName
        singleItemView.setItemCategory = individualData.itemCategory
        singleItemView.setItemID = individualData.ID
        singleItemView.setItemPrice = NSNumber(value: individualData.price)
        singleItemView.setItemBrand = individualData.brand
        singleItemView.setItemQuality = individualData.quality
        
        navigationController?.pushViewController(singleItemView, animated: true)
    }
    
    func showPrice(_ price: Double) -> String {
        return "$\(price)"
    }
}
